flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java

```diff
@@ -178,14 +185,20 @@ public void notifyCheckpointComplete(long checkpointId) throws Exception {
     }
   }
 
-  private void commitUpToCheckpoint(NavigableMap<Long, List<DataFile>> dataFilesMap,
+  private void commitUpToCheckpoint(NavigableMap<Long, Byte[]> manifestsMap,
                                     String newFlinkJobId,
-                                    long checkpointId) {
-    NavigableMap<Long, List<DataFile>> pendingFileMap = dataFilesMap.headMap(checkpointId, true);
+                                    long checkpointId) throws IOException {
+    NavigableMap<Long, Byte[]> pendingManifestMap = manifestsMap.headMap(checkpointId, true);
+
+    List<ManifestFile> manifestFiles = Lists.newArrayList();
+    for (Byte[] manifestData : pendingManifestMap.values()) {
+      ManifestFile manifestFile = ManifestFiles.decode(ArrayUtils.toPrimitive(manifestData));
+      manifestFiles.add(manifestFile);
+    }
 
     List<DataFile> pendingDataFiles = Lists.newArrayList();
-    for (List<DataFile> dataFiles : pendingFileMap.values()) {
-      pendingDataFiles.addAll(dataFiles);
+    for (ManifestFile manifestFile : manifestFiles) {
+      pendingDataFiles.addAll(FlinkManifest.read(manifestFile, table.io()));
```

Comment:
I think it is okay to do this, but the append operation can append whole manifest files and either copy the manifest contents into an appropriate path for you, or take ownership of the manifest and manage it so it doesn't need to be rewritten. You might consider using `AppendFiles.appendManifest` instead.

If you do choose to append manifests, then we would want to know whether the manifests are written into the table's metadata space (path came from `TableOperations.metadataFilePath`) or whether the manifest was written to a temp location, like @stevenzwu's suggestion. In the latter case, we would want to request that Iceberg rewrite the manifests in the commit operation.

Reply:
I read all data files from the written manifests, because there may be one of the operations:  `ReplacePartitions` or `AppendFiles`.  For `AppendFiles` , it's true that we could just use `appendManifest` if it's in the metadata location, while for `ReplacePartitions`, it does not have an `appendManifest` so we still have to read all DataFiles and call `ReplacePartitions#addFile` one by one.  So here I use the unified method to handle both cases.

Reply:
should we add the `appendManifest` to `ReplacePartitions` too?

Reply:
I think you concern is :  the manifests content is so large in an extreme case  that it will take much resources to copy those entries ?  If we really want to save the copy, we will need : 
1.   Set the `compatibility.snapshot-id-inheritance.enabled` to be `true`,  its default value is `false` and in that case it will still copy the contents to assign the correct snapshot id.  https://github.com/apache/iceberg/blob/dc661d417098265de61383650e784d1b39a57209/core/src/main/java/org/apache/iceberg/FastAppend.java#L105 . 
2.  Add the necessary `appendManifest` interface and implementation for `ReplacePartition`. 

I agree that it's nice to have that improvement, but as the release time point is coming, I'd rather not to block this the next release 0.10.0, this should not break the user's experience or compatibility.

Reply:
We would need to scan the manifest to recover the partitions that need to be replaced. We can definitely add that, but I agree that we don't need to block here.